File: src/main/java/club/xiaojiawei/listener/AbstractLogListener.java
package club.xiaojiawei.listener;

import club.xiaojiawei.custom.LogRunnable;
import club.xiaojiawei.data.ScriptStaticData;
import club.xiaojiawei.utils.SystemUtil;
import javafx.beans.property.BooleanProperty;
import lombok.Getter;
import lombok.Setter;
import lombok.extern.slf4j.Slf4j;

import javax.annotation.Resource;
import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.io.RandomAccessFile;
import java.util.concurrent.ScheduledFuture;
import java.util.concurrent.ScheduledThreadPoolExecutor;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicReference;

/**
 * @author 肖嘉威
 * @date 2023/9/20 16:54
 * @msg
 */
@Slf4j
public abstract class AbstractLogListener {
    @Resource
    protected ScheduledThreadPoolExecutor listenFileThreadPool;
    @Resource
    protected AtomicReference<BooleanProperty> isPause;
    @Setter
    protected static File logDir;
    @Getter
    protected RandomAccessFile accessFile;
    protected ScheduledFuture<?> logScheduledFuture;
    protected String logName;
    protected long listenInitialDelay;
    protected long listenPeriod;
    protected TimeUnit listenUnit;

    public AbstractLogListener(String logName,
                               long listenInitialDelay,
                               long listenPeriod,
                               TimeUnit listenUnit) {
        this.logName = logName;
        this.listenInitialDelay = listenInitialDelay;
        this.listenPeriod = listenPeriod;
        this.listenUnit = listenUnit;
    }

    protected abstract void readOldLog();
    protected abstract void listenLog();
    protected void otherListen(){};
    protected void cancelOtherListener(){};
    public synchronized void listen(){
        if (logScheduledFuture != null && !logScheduledFuture.isDone()){
            log.warn(logName + "正在被监听，无法再次被监听");
            return;
        }
        File logFile = createFile();
        closeLogStream();
        log.info("开始监听" + logName);
        try {
            accessFile = new RandomAccessFile(logFile, "r");
            readOldLog();
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
        logScheduledFuture = listenFileThreadPool.scheduleAtFixedRate(new LogRunnable(this::listenLog), listenInitialDelay, listenPeriod, listenUnit);
        otherListen();
    }
    private File createFile(){
        File screenLogFile = new File(logDir.getAbsolutePath() + "/" + logName);
        if (!screenLogFile.exists()){
            try(FileWriter fileWriter = new FileWriter(screenLogFile)){
                fileWriter.write("#created by " + ScriptStaticData.SCRIPT_NAME);
            } catch (IOException e) {
                throw new RuntimeException(e);
            }
        }
        return screenLogFile;
    }
    private void closeLogStream(){
        if (accessFile != null){
            try {
                accessFile.close();
                accessFile = null;
            } catch (IOException e) {
                throw new RuntimeException(e);
            }
        }
    }
    public void cancelListener(){
        if (logScheduledFuture != null && !logScheduledFuture.isDone()){
            logScheduledFuture.cancel(true);
            log.info("已停止监听" + logName);
            SystemUtil.delay(2000);
        }
        cancelOtherListener();
    }
}

File: hs-script-app/src/main/java/club/xiaojiawei/hsscript/listener/log/AbstractLogListener.kt
package club.xiaojiawei.hsscript.listener.log

import club.xiaojiawei.hsscriptbase.config.LISTEN_LOG_THREAD_POOL
import club.xiaojiawei.hsscriptbase.config.log
import club.xiaojiawei.hsscript.interfaces.closer.ScheduledCloser
import club.xiaojiawei.hsscript.listener.WorkTimeListener
import club.xiaojiawei.hsscript.status.PauseStatus
import club.xiaojiawei.hsscript.status.TaskManager
import club.xiaojiawei.hsscriptbase.util.isFalse
import java.io.File
import java.io.FileWriter
import java.io.RandomAccessFile
import java.util.concurrent.ScheduledFuture
import java.util.concurrent.TimeUnit

/**
 * 在 [LogListenerConfig.logListener] 添加新的LogListener
 * @author 肖嘉威
 * @date 2023/9/20 16:54
 */
abstract class AbstractLogListener(
    protected var logFileName: String,
    protected var listenInitialDelay: Long,
    protected var listenPeriod: Long,
    protected var listenTimeUnit: TimeUnit
) : ScheduledCloser {

    init {
        TaskManager.addTask(this)
    }

    protected var innerLogFile: RandomAccessFile? = null

    val logFile: RandomAccessFile?
        get() = innerLogFile

    var logFilePath: String? = null

    private var logScheduledFuture: ScheduledFuture<*>? = null

    var nextLogListener: AbstractLogListener? = null

    fun setNextLogListener(nextLogListener: AbstractLogListener): AbstractLogListener {
        return nextLogListener.also { this.nextLogListener = it }
    }

    protected abstract fun dealOldLog()

    protected abstract fun dealNewLog()

    private fun listenNextListener() {
        nextLogListener?.listen()
    }

    fun listen() {
        synchronized(this){
            logScheduledFuture?.let {
                if (!it.isDone) {
                    log.warn { logFileName + "正在被监听，无法再次被监听" }
                    listenNextListener()
                    return
                }
            }
            closeLogFile()
            val logFile = createLogFile()
            logFile ?: let {
                log.error { logFileName + "初始化失败" }
                return
            }
            log.info { "开始监听日志$logFileName" }
            try {
                this.innerLogFile = RandomAccessFile(logFile, "r")
                dealOldLog()
            } catch (e: Exception) {
                log.error(e) {}
                return
            }
            logScheduledFuture = LISTEN_LOG_THREAD_POOL.scheduleWithFixedDelay({
                if (PauseStatus.isPause || !WorkTimeListener.working) {
                    stopAll()
                } else {
                    try {
                        dealNewLog()
                    } catch (e: InterruptedException) {
                        log.warn(e) { logFileName + "监听中断" }
                    } catch (e: Exception) {
                        log.error(e) { logFileName + "监听发生错误" }
                    }
                }
            }, listenInitialDelay, listenPeriod, listenTimeUnit)
            listenNextListener()
        }
    }

    private fun createLogFile(): File? {
        return logPath?.let {
            val logFile = it.resolve(logFileName)
            logFile.exists().isFalse {
                FileWriter(logFile).use { fileWriter ->
                    fileWriter.write("")
                }
            }
            logFilePath = logFile.absolutePath
            logFile
        }
    }

    private fun closeLogFile() {
        synchronized(this) {
            innerLogFile?.let {
                it.close()
                innerLogFile = null
                logFilePath = null
            }
        }
    }

    private fun closeLogListener() {
        synchronized(this) {
            logScheduledFuture?.let {
                it.isDone.isFalse {
                    it.cancel(true)
                }
            }
        }
    }

    override fun stopAll() {
        closeLogListener()
    }

    companion object {
        var logPath: File? = null
    }

}
